test/node-report/testcfg.py

```diff
@@ -0,0 +1,6 @@
+import sys, os
+sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
+import testpy
+
+def GetConfiguration(context, root):
+  return testpy.SimpleTestConfiguration(context, root, 'node-report')
```

Comment:
Can these tests be run in parallel? If yes, that would be preferable (`ParallelTestConfiguration`)

Reply:
the issue is with the generated reports. If you see the logic in helper.js that finds and enumerates the reports, I believe it can lead to race and unexpected result if we run those in parallel.

Are there a strong reason why tests should be run in parallel? reduction in test time? or some thing else?

Reply:
@gireeshpunathil Are the reports not written to the test's tmpdir? (See the trace-events tests for examples). That would also take care of clean up.

Reply:
@richardlau - no, by default report is generated in the `cwd` of the process; configurable with `setDiagnosticReportOptions` to point to anywhere. The issue is that report generation location is part of the report's function itself, and we have tests and negative tests to validate those. So in summary I guess if we use `tmpdir` to favor of tests' sanity, we might be ignoring the functional correctness of report?

Reply:
@gireeshpunathil Almost all of the tests exec/fork/spawn a child process. If we pass `cwd: common.tmpDir` option to set the cwd of the child process then by default the reports will go there. 

See, for example, https://github.com/nodejs/node/blob/23f8b7d334db86a78464f8694c015afab6dedfc3/test/parallel/test-trace-events-process-exit.js#L1-L23 which changes the cwd for a trace-events test -- the trace file ends up in the changed cwd.

Reply:
thanks, makes sense - sadly, there are other tests too where the parent themselves produce reports!

let me think over it - probably it makes sense to run all the tests as children processes - to manages the report files?

Reply:
@richardlau, @addaleax  - we played around with this a bit. As the parent does not have a way of figuring out the name of the report that the child generates, it employs a function called `findReports`, basically queries the whole folder for files matching a pattern. When we parallelize tests and use a temp folder for that, the issue just shifts from test folder to temp folder, but does not go away - test A can process test B's report!

potential solutions are either of:
 - create unique temp folders for each test, leveraging `tmpdir` module and some hacks
 - instead of querying for reports, child communicate the report name to the parent (somehow)

Before attempting these, I would like to know: what benefit we gain through running things in parallel? overall test running time, an attempt to simulate stress on node? or something else?

Reply:
> test A can process test B's report

@gireeshpunathil That shouldn't happen. The test harness should create a new temp dir for each test it runs and `tmpdir.refresh()` should clear it.

Reply:
$  cat test.js
```js
const tmpdir = require('./common/tmpdir');
const path = require('path');
tmpdir.refresh();
const file = path.resolve(path.join(tmpdir.path, 'report.json'));
console.log(file);
setTimeout(()=> {}, 10000);
```
```
$ node test.js &
[1] 23030
$  /home/gireesh/nr/o/node/test/.tmp/report.json

$  ../node test.js &
[2] 23032
$  /home/gireesh/nr/o/node/test/.tmp/report.json
```
As you can see, both the tests are using the same folder!

Reply:
thanks, that is what I also thought; but does not look like it works as intended:

cd test/foo
ls 
```
testcfg.py   test-foo10.js  test-foo2.js  test-foo4.js	test-foo6.js  test-foo8.js  test-foo.js
testcfg.pyc  test-foo1.js   test-foo3.js  test-foo5.js	test-foo7.js  test-foo9.js
```
cat test-foo.js
```
const tmpdir = require('../common/tmpdir');
const path = require('path');
tmpdir.refresh();
const file = path.resolve(path.join(tmpdir.path, 'report.json'));
console.error(file);
require('assert').ok(null);
```
all test files are identical.
cat testcfg.py
```
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
import testpy

def GetConfiguration(context, root):
  return testpy.ParallelTestConfiguration(context, root, 'foo')
```
cd ../../
tools/test.py foo
among all the error messges, the file name that gets printed is:
`/home/gireesh/nr/o/node/test/.tmp.0/report.json` , for all 10 tests.

Reply:
You need to pass `-J` to specify the number of tasks to run in parallel (it defaults to 1): https://github.com/nodejs/node/blob/11c17e4bc912af7dd5b6ef47176fcccee9064b50/tools/test.py#L1346-L1347

Reply:
thanks, that clarifies, we will try with this approach.

Reply:
Done parallelizing the test cases. For achieving this purpose, I have implemented the report generating logic in the child.

Reply:
I think you still need to return `ParallelTestConfiguration` instead of `SimpleTestConfiguration` here.

Reply:
Done. I was not aware of this functionality. Thanks for pointing out.